pkg/master/reconcilers/lease.go

```diff
@@ -205,9 +207,9 @@ func (r *leaseEndpointReconciler) doReconcile(serviceName string, endpointPorts
 
 	if !formatCorrect || !ipCorrect {
 		// repopulate the addresses according to the expected IPs from etcd
-		e.Subsets[0].Addresses = make([]corev1.EndpointAddress, len(masterIPs))
-		for ind, ip := range masterIPs {
-			e.Subsets[0].Addresses[ind] = corev1.EndpointAddress{IP: ip}
```

Comment:
does this mean that skewed masters will fight over the content of the kubernetes service endpoints (1.14-level ones will strip the nodeName, 1.15-level ones will try to add it)?

Reply:
Yes, you are right. Hence, we should backport this fix to the all supported releases

Reply:
We cannot require the latest patch release for skew compatibility

Reply:
Could you, please, clarify why? At this moment, we have a bug that causes to an inconsistent state of the k8s default service endpoint - every others endpoints have nodeName, but the k8s default service one doesn't. According to the semver, patch release allows us to add bug fixes. During cluster upgrade, it is acceptable to have different versions, because this patch does not affect the cluster and everything will work fine.

Reply:
> At this moment, we have a bug that causes to an inconsistent state of the k8s default service endpoint - every others endpoints have nodeName, but the k8s default service one doesn't.

That's not a bug. endpoint address nodeName is optional, and the kubernetes service endpoint creation code opts not to populate it currently.

If we want to start populating it in a compatible way where it makes sense, that could certainly be a useful enhancement, but we *must* do so in a way that does not cause problems with our [supported API server skew levels](https://kubernetes.io/docs/setup/version-skew-policy/).

> During cluster upgrade, it is acceptable to have different versions, because this patch does not affect the cluster and everything will work fine.

If I'm reading this PR correctly, it would make a 1.14.0 API server and an API server with this change duel with each other over the content of the kubernetes service endpoints. Backporting changes to 1.14.x does not allow us to add features in 1.15.x that do not work with 1.14.0.

Reply:
Thanks for the clarification. I need time to think about how we can make this in compatible way. Any proposals are appreciated :)

Reply:
@liggitt I've commited changes (the last commit) that store nodeName in compatible way. For now, every apiserver tracks only its own nodeName and if we have different versions in the cluster, the worst case is endpoint addresses will be with nodeName (apiserver 1.15.x) and without it (older versions) simultaneously.